README.md
swift-switch-account
====================

`swift-switch-account` (abbreviated sw2) is a client-server
application that allows individuals to set Swift environment variables
used by `python-swiftclient` for accessing Swift object stores.

## Basic Operations

The sw2 client will prompt the user for their directory password and
the Swift account they wish to access.  Given successful
authentication of the credentials and authorization to the account,
the client will set the environment variables:

- `ST_USER`
- `ST_KEY`

with values appropriate for the Swift account.  If the client is
instructed to persist these environment variables, the client will
write a file called `.swiftrc` (or `.swift.cshrc` for C shell
variants) that can be read in by the shell.

Note that `ST_AUTH` (with the Swift authorization URL) is not set by
either this script nor is it written to a swiftrc file.

If either authorization or authentication fail, the user is notified
with an appropriate error message indicating the cause.  Existing
Swift environment variables and the `.swiftrc` file are left as-is.

## Use

`sw2account [--server-url=<url>] [--persist] <account>`

Put appropriate Swift credentials into current user's environment for
Swift account `account`.  If `--persist` is specified, Swift
credentials will be written into a file in the user's home directory.

# Install and Setup

## Shell Environment for sw2account

The environment variable `SW2_AUTH` indicates the root URL where the
script can find the sw2server process. 

Since the script must update the current shell environment, it's
necessary to configure either the user's login or the global
(/etc/profile\*) startup files. Thus, a function needs to be defined
that reads the output of the `sw2account.py` command into the current
environment via `eval`:

```bash
sw2account () {
 eval `/path/to/sw2account.py bash $*; `
}
```

How to implement this is site and shell specific.  For bash, if a site
wanted these set for all users logging into a system, you could create
a file `/etc/profile.d/sw2account` which contained both the definition
of that function and an appropriate value for `SW2_AUTH`.

## Configuring sw2 server components

Retreive the code into a suitable location (`/var/www/sw2srv` for
example).  Configure the application from the template- it's basic
python, most parameters are fairly self-explanatory.

`keyfile` is important: set this to the location of the file
containing the Swift certificates.  Set `key_fields` to indicate the
order of the fields in the csv file.

If you want to use dummy certificates, the Makefile in the `certs`
subdirectory contains the necessary commands to make some self-signed
certs suitable for testing.

## Running the flask server

For testing, the built-in flask server is suitable.  The server can be
started using the "swift-switch-server.py" script.  Make sure that
values in the configuration file are appropriate for your needs.

## Running sw2srv with Apache, SSL, and WSGI:

This is a fairly common stack.  Below is a sample vhost configuration
for Apache:

```
<VirtualHost *:443>
  ServerName sw2srv.fhcrc.org

  ## Vhost docroot
  DocumentRoot "/var/www/sw2srv"

  <Directory "/var/www/sw2srv">
    Options Indexes FollowSymLinks MultiViews
    AllowOverride None
    Require all granted
  </Directory>

  ## Logging
  ErrorLog "/var/log/apache2/sw2srv.foo.org_error_ssl.log"
  ServerSignature Off
  CustomLog "/var/log/apache2/sw2srv.foo.org_access_ssl.log"
combined

  ## SSL directives
  SSLEngine on
  SSLCertificateFile      "/etc/ssl/certs/cert.pem"
  SSLCertificateKeyFile   "/etc/ssl/private/cert.key"
  SSLCACertificatePath    "/etc/ssl/certs"
  WSGIApplicationGroup ${GLOBAL}
  WSGIDaemonProcess sw2srv processes=1 threads=2
  WSGIProcessGroup sw2srv
  WSGIScriptAlias /sw2srv "/var/www/sw2srv/sw2srv.wsgi"
  WSGIPassAuthorization On
</VirtualHost>
```

Note that WSGIPassAuthorization is necessary for this application to
get the authentication passed through from Apache.  Other WSGI
parameters may require adjustments for your environment.

The WSGI script needs to be updated with the installation location:

```
import sys
sys.path.insert(0, '/var/www/sw2srv/sw2srv' )
from sw2srv import server as application
```

Update the inserted path to the path where the application init is
found.

# Internals

## How it Works

The client is run with an argument indicating the Swift account the
user wishes to access.  The client then prompts for the password of the
currently logged-in user account and bundles this information into an
encrypted REST request to the server.  Authentication credentials are
provided to the server as basic HTTP authentication to a URL
containing the Swift account name.

    https://user:password@toolbox.fhcrc.org/swift/account/foo_b

If the credential pass authentication (to the directory) the server
then parses this request.  The authorization group is built by
appending `_grp` to the account name (e.g. `foo_b_grp`).  The server
checks authorization (see [Authorization Logic](#authorization-logic))
for this group and if successful, looks up the Swift account
credentials stored in a text file on the server.  These credentials
(the Swift auth URL, user account, and S3 key) are then encoded as a
JSON response to the client with an HTTP status code of 200.

Minimally the client will set Swift environment variables for the
current session:

- `ST_USER`
- `ST_KEY`

Setting these environment variables is done via a bit of sleight-of-hand
stolen from Modules.  The user must source the init file appropriate for
her shell which installs `sw2account` as a function.  This function
uses eval to run the script which produces bash-like functions as its
output.

If instructed, the client will write/over-write a file in the current
user's home directory.

## Errors:

### Bad login credentials:

- server returns 401 and empty response
- client indicates login failure on terminal and non-zero exit

### User not a member of appropriate groups

- server returns 403 and empty response
- client indicates that user does not have access to group
  and sets non-zero exit code.

### Swift Group/Account does not exist (i.e. no key)

- server returns 404 and JSON encoded error message
- client relays error message to terminal and non-zero exit

### LDAP connection errors

- server returns 500
- client indicates service unavailable

## Authorization Logic

Authorization for an account is controlled by a user's membership in
an eponymous group or in the eponymous group's `managedBy` entity.

If a user requests access to the Swift account `foo_b`, the server
first checks to see if a group named `foo_b_grp` is in the user's
`memberOf` list.  Next, the user is checked to see if the user is a
member of the group managing the group `foo_b_grp` using the same
`memberOf` check.

If either check succeeds, the credentials for the requested account
are returned.


/* vim: set textwidth=70 : */


debian/changelog
swift-switch-account (0.2.1) trusty; urgency=low

    Added additonal logic to search for a group ending `_swift_grp`

 -- Dirk Petersen <petersen@fredhutch.org> Wed, 08 Jun 2016 10:53:01 -0700

swift-switch-account (0.2.0-4) trusty; urgency=low

    removed prompting for passwords

 -- Michael Gutteridge <mrg@fredhutch.org> Wed, 02 Sep 2015 15:44:53 -0700

swift-switch-account (0.2.0-3) trusty; urgency=low

    added --[no]-save-password option

 -- Michael Gutteridge <mrg@fredhutch.org> Wed, 19 Aug 2015 15:30:24 -0700

swift-switch-account (0.2.0-2) trusty; urgency=low

    bug fixes

 -- Michael Gutteridge <mrg@fredhutch.org> Tue, 04 Aug 2015 06:24:39 -0700

swift-switch-account (0.2.0-1) trusty; urgency=low

  Adding Swiftstack V2 authentication capabilities

 -- Michael Gutteridge <mrg@fredhutch.org> Mon, 27 Jul 2015 11:24:05 -0700

swift-switch-account (0.1.2-1) trusty; urgency=low

  * added zsh support

 -- Michael Gutteridge <mrg@fredhutch.org>  Thu, 15 Jan 2015 09:34:18 -0800

swift-switch-account (0.1.1-1) trusty; urgency=low

  * Added version argument
  * Release to produciton

 -- Michael Gutteridge <mrg@fredhutch.org>  Wed, 14 Jan 2015 08:50:49 -0800

swift-switch-account (0.1.0-1) trusty; urgency=low

  * Initial release 

 -- Michael Gutteridge <mrg@fredhutch.org>  Tue, 06 Jan 2015 10:40:48 -0800


sw2account/sw2account.py
#!/usr/bin/python

__version__='0.2.1'

import argparse
import logging
import os
import os.path
import sys
import requests
import getpass
import ConfigParser

def _persist( export, rcfile ):
    f = open( rcfile, 'w' )
    logging.debug( "writing to {}".format( rcfile ) )
    f.write( "\n".join( export ) )
    f.close()
    os.chmod( rcfile, 0600 )
    logging.info( "saved Swift credentials" )

def sh(creds, auth_version, savepw=False, persist=False ):
    export = []
    if auth_version == 'v1':
        export.append(
            "unset OS_USERNAME OS_PASSWORD OS_TENANT_NAME OS_AUTH_URL" )
        export.append(
            "unset OS_AUTH_TOKEN OS_STORAGE_URL" )
        export.append( "export ST_USER='{}'".format( creds['account'] ) )
        export.append( "export ST_KEY='{}'".format( creds['password'] ) )
        export.append( "export ST_AUTH='{}'".format( v1AuthUrl ) )
    else:
        export.append(
            "unset ST_USER ST_KEY ST_AUTH" )
        export.append( "export OS_USERNAME='{}'".format( creds['user'] ) )
        export.append(
            "export OS_TENANT_NAME='AUTH_Swift_{}'".format( creds['account'] ) )
        if savepw:
            export.append(
                "export OS_PASSWORD='{}'".format( creds['password'] ) )
        export.append( "export OS_AUTH_URL='{}'".format( v2AuthUrl ) )

    print ";".join( export )

    if persist:
        rcfile = os.environ[ 'HOME' ] + "/.swiftrc"
        logging.debug( "persisting environment variables" )
        _persist( export, rcfile )

def csh(creds, auth_version, savepw=False, persist=False ):
    export = []
    if auth_version == 'v1':
        export.append(
            "unsetenv OS_USERNAME OS_PASSWORD OS_TENANT_NAME OS_AUTH_URL" )
        export.append(
            "unsetenv OS_AUTH_TOKEN OS_STORAGE_URL" )
        export.append( "setenv ST_USER '{}'".format( creds['account'] ) )
        export.append( "setenv ST_KEY '{}'".format( creds['password'] ) )
        export.append( "setenv ST_AUTH '{}'".format( v1AuthUrl ) )
    else:
        export.append(
            "unsetenv ST_USER ST_KEY ST_AUTH" )
        export.append( "setenv OS_USERNAME '{}'".format( creds['user'] ) )
        export.append(
            "setenv OS_TENANT_NAME 'AUTH_Swift_{}'".format( creds['account'] ) )
        if savepw:
            export.append(
                "setenv OS_PASSWORD '{}'".format( creds['password'] ) )
        export.append( "setenv OS_AUTH_URL '{}'".format( v2AuthUrl ) )

    print ";".join( export )

    if persist:
        rcfile = os.environ[ 'HOME' ] + "/.swift.cshrc"
        logging.debug( "persisting environment variables" )
        _persist( export, rcfile )

shell_output = {
    'sh': sh,
    'ksh': sh,
    'bash': sh,
    'zsh': sh,
    'csh': csh,
    'tcsh': csh
}

class LocalParser( argparse.ArgumentParser ):
    def error( self, message ):
        sys.stderr.write( "Error: too few arguments\n" )
        sys.stderr.write( "usage: sw2account lastname_f\n" )
        sys.stderr.write(
           "use \"sw2account --help\" for full help information\n" )
        sys.exit(1)

    def print_help( self ):
        self._print_message( self.format_help(), sys.stderr )
        sys.exit(0)

def return_v1_auth( args ):
    # If server URL is unspecified, look for "SW2_URL" in current environment
    account = args.account
    server_url = args.server_url

    logging.debug(
        'asking {} for credentials for {}'.format( server_url, account )
    )

    if not server_url:
        try:
            server_url = os.environ[ 'SW2_URL' ]
        except KeyError:
            logging.error( "Server URL is unset (not in arguments or SW2_URL)" )
            sys.exit(1)

    # Add account name to URL
    server_url = '/'.join( [ server_url, account ] )
    logging.debug( 'final url is {}'.format( server_url ) )

    # Get user authentication credentials
    user = getpass.getuser()
    passwd = getpass.getpass( 'Enter password for {}: '.format(user) )

    # Get account credentials from server_url
    r = requests.get( server_url, verify = args.verify_ssl, auth=( user, passwd ) )
    if r.status_code == 200:
        creds = r.json()
        logging.debug(
            "got credentials for account {}".format( creds['account'] )
        )

        creds['url'] = 'https://tin/some/crap'
        shell_output[ args.shell ](
            creds=creds,
            persist=args.persist,
            auth_version=args.auth_version
        )

    elif r.status_code == 401:
        logging.error(
            "invalid username/password supplied to server"
        )
    elif r.status_code == 403:
        logging.error(
            "user {} is not permitted to use {} ({})".format(
                user, account, r.status_code
            )
        )
    elif r.status_code == 404:
        try:
            message = r.json()['message']
        except KeyError:
            logging.error( "404 returned from server with no message" )
            sys.exit(1)
        logging.error("{} (HTTP{})".format(
                 message, r.status_code
            )
        )
    else:
        logging.error(
            "error {} retrieving credentials from server".format(
                r.status_code
            )
        )

def return_v2_auth( args ):
    creds = {}
    # authentication is done using Swiftstack version 2 authentication

    # requires additional "tenant name" in addition to username and password
    creds['account'] = args.account

    # take username password from currently logged in user
    creds['user'] = getpass.getuser()
    if args.savepw:
        logging.warning( "Saving passwords is insecure and not recommended." )
        creds['password'] = getpass.getpass(
            'Enter password for {}: '.format( creds['user'] ) )
    logging.debug(
        "got credentials for account {}".format( creds['account'] )
    )
    if args.savepw:
        logging.debug( 'saving password in rc and environment' )

    shell_output[ args.shell ](
        creds=creds,
        persist=args.persist,
        savepw=args.savepw,
        auth_version=args.auth_version
    )

def add_common_args( aparser ):
    aparser.add_argument(
        'shell',
        help = "format output for shell <shell>",
        choices = shell_output.keys()
    )
    aparser.add_argument(
        'account',
        help = "retrieve credentials for account <account>"
    )
    aparser.add_argument(
        '--config',
        default = "/etc/sw2account.cfg",
        help = "configuration file to use (default=/etc/sw2account.cfg)"
    )
    aparser.add_argument(
        '--stack',
        default = "default",
        help = "stack name to authentication against (see configfile)"
    )
    aparser.add_argument(
        '--save', '--persist',
        dest = 'persist',
        action = 'store_true',
        help = "write credentials to $HOME/.swiftrc"
    )
    aparser.add_argument(
        '--no-save', '--no-persist',
        dest = 'persist',
        action = 'store_false',
        help = "do not write credentials to $HOME/.swiftrc"
    )
    aparser.add_argument(
        '--version', '-v',
        help = "show script version",
        action = 'version',
        version = "sw2account version {}".format( __version__)
    )
    aparser.add_argument(
        '--debug',
        action = "store_true",
        help = "log level for client"
    )

if __name__ == "__main__":

    # Get the config first
    # Need to prime the pump to find defaults
    tparse = argparse.ArgumentParser()
    tparse.add_argument(
        '--config',
        default = "/etc/sw2account.cfg",
        help = "configuration file to use (default=/etc/sw2account.cfg)"
    )
    tparse.add_argument(
        '--stack',
        default = "default",
        help = "stack name to authenticate against (see configfile)"
    )
    args, unknown = tparse.parse_known_args()

    # Read config file with defaults
    if not os.path.isfile( args.config ):
        logging.error( "missing config file %s", args.config )
        sys.exit(1)

    appdefaults = ConfigParser.ConfigParser()
    try:
        appdefaults.read( args.config )
        logging.debug( "reading config from %s", args.config )
    except ConfigParser.ParsingError:
        logging.error(
            "error reading configuration file %s - check format", args.config
        )
        sys.exit(1)

    try:
        v1AuthUrl = appdefaults.get( args.stack, 'v1AuthUrl' )
        v2AuthUrl = appdefaults.get( args.stack, 'v2AuthUrl' )
        auth_version_default = appdefaults.get(
            args.stack, 'auth_version_default' )
    except ConfigParser.NoSectionError:
        logging.error( "Stack '%s' not configured in configfile %s",
                      args.stack, args.config )
        sys.exit(1)
    except ConfigParser.NoOptionError:
        logging.error(
            "Configfile %s does not contain correct entries for stack '%s'",
            args.config, args.stack
        )
        sys.exit(1)


    # Fix argument order so that v1/v2 is first argument
    try:
        if sys.argv[1] not in ['v1','v2'] and (
            '-h' not in sys.argv or '--help' not in sys.argv ):
            if 'v1' in sys.argv:
                logging.debug( "reordering arguments to put v1 arg at head" )
                sys.argv.remove('v1')
                sys.argv.insert(1, 'v1')
            elif 'v2' in sys.argv:
                logging.debug( "reordering arguments to put v2 arg at head" )
                sys.argv.remove('v2')
                sys.argv.insert(1, 'v2')
            else:
                logging.debug( "setting default version" )
                sys.argv.insert(1, auth_version_default)
    except IndexError:
        pass

    #parser = LocalParser()
    parser = argparse.ArgumentParser()
    subparser = parser.add_subparsers(
        dest = "auth_version", help='authentication version to use'
    )

    v1parser = subparser.add_parser('v1', help='use version 1 authentication')
    add_common_args( v1parser )
    v1parser.add_argument(
        '--server-url',
        help = "URL of server with account data"
    )
    v1parser.add_argument(
        '--no-verify-ssl',
        action = "store_false",
        dest = "verify_ssl",
        help = "verify ssl certification",
    )
    v1parser.add_argument(
        '--verify-ssl',
        action = "store_true",
        dest = "verify_ssl",
        help = "verify ssl certification",
    )

    v2parser = subparser.add_parser('v2', help='use version 2 authentication')
    add_common_args( v2parser )

    v2parser.add_argument(
        '--save-password',
        action = "store_true",
        dest = "savepw",
        help = "save password in rc and environment- not recommended",
    )

    v2parser.add_argument(
        '--no-save-password',
        action = "store_false",
        dest = "savepw",
        help = "(default) do not save password in rc and environment- recommended",
    )

    args = parser.parse_args()

    if args.debug:
        logging.basicConfig( level=logging.DEBUG )
    logging.debug( 'arguments: %s', args )

    if args.shell in [ 'bash', 'ksh', 'sh', 'zsh' ]:
        rcfile = os.environ[ 'HOME' ] + "/.swiftrc"
    elif args.shell in [ 'tcsh', 'csh' ]:
        rcfile = os.environ[ 'HOME' ] + "/.swift.cshrc"

    logging.debug( 'checking for rcfile: {}'.format(rcfile))
    if not os.path.isfile( rcfile ):
        parser.set_defaults( persist=True )
        logging.debug( "no rcfile: set persist default to true" )
    else:
        parser.set_defaults( persist=False )
        logging.debug( "found rcfile: set persist default to false" )

    args = parser.parse_args()
    logging.debug( "persist is set to {} after second wash".format(
        args.persist
    ))

    if args.auth_version == 'v1':
        return_v1_auth( args )
    elif args.auth_version == 'v2':
        return_v2_auth( args )


